File: rustRaytracer/src/material.rs
use crate::{
    hittable::HitRecord,
    ray::Ray,
    vec3::{dot, random_unit_vector, reflect, refract, Color},
};
use rand::Rng;

pub trait Material: Send + Sync {
    fn scatter(&self, r_in: &Ray, rec: &HitRecord) -> (bool, Color, Ray);
}

pub struct Lambertian {
    albedo: Color,
}

impl Lambertian {
    pub fn new(albedo: Color) -> Self {
        Self { albedo }
    }
}

impl Material for Lambertian {
    fn scatter(&self, _r_in: &Ray, rec: &HitRecord) -> (bool, Color, Ray) {
        let mut scatter_direction = rec.normal + random_unit_vector();
        
        // Catch degenerate scatter direction
        if scatter_direction.near_zero() {
            scatter_direction = rec.normal;
        }
        
        let scattered = Ray::new(rec.p, scatter_direction);
        (true, self.albedo, scattered)
    }
}

pub struct Metal {
    albedo: Color,
    fuzz: f64,
}

impl Metal {
    pub fn new(albedo: Color, fuzz: f64) -> Self {
        Self {
            albedo,
            fuzz: if fuzz < 1.0 { fuzz } else { 1.0 },
        }
    }
}

impl Material for Metal {
    fn scatter(&self, r_in: &Ray, rec: &HitRecord) -> (bool, Color, Ray) {
        let reflected = reflect(&r_in.direction(), &rec.normal);
        let scattered = Ray::new(
            rec.p,
            reflected + random_unit_vector() * self.fuzz,
        );
        let scatter_happened = dot(&scattered.direction(), &rec.normal) > 0.0;
        (scatter_happened, self.albedo, scattered)
    }
}

pub struct Dielectric {
    ir: f64, // Index of refraction
}

impl Dielectric {
    pub fn new(index_of_refraction: f64) -> Self {
        Self {
            ir: index_of_refraction,
        }
    }

    fn reflectance(&self, cosine: f64, ref_idx: f64) -> f64 {
        // Use Schlick's approximation for reflectance
        let mut r0 = (1.0 - ref_idx) / (1.0 + ref_idx);
        r0 = r0 * r0;
        r0 + (1.0 - r0) * (1.0 - cosine).powi(5)
    }
}

impl Material for Dielectric {
    fn scatter(&self, r_in: &Ray, rec: &HitRecord) -> (bool, Color, Ray) {
        let attenuation = Color::new(1.0, 1.0, 1.0);
        let refraction_ratio = if rec.front_face { 1.0 / self.ir } else { self.ir };

        let unit_direction = r_in.direction().length().recip() * r_in.direction();
        let cos_theta = f64::min(dot(&-unit_direction, &rec.normal), 1.0);
        let sin_theta = (1.0 - cos_theta * cos_theta).sqrt();

        let cannot_refract = refraction_ratio * sin_theta > 1.0;
        let mut rng = rand::thread_rng();
        let will_reflect = self.reflectance(cos_theta, refraction_ratio) > rng.gen::<f64>();

        let direction = if cannot_refract || will_reflect {
            reflect(&unit_direction, &rec.normal)
        } else {
            refract(&unit_direction, &rec.normal, refraction_ratio)
        };

        let scattered = Ray::new(rec.p, direction);
        (true, attenuation, scattered)
    }
}

File: python-Raytracer/material.py
import random
import math
from ray import Ray
from vec3 import dot, random_unit_vector, reflect, refract, unit_vector, Color
from hittable import HitRecord

class Material:
    def scatter(self, r_in: Ray, rec: HitRecord) -> tuple[bool, Color, Ray]:
        """Returns (scatter_happened, attenuation, scattered_ray)"""
        pass


class Lambertian(Material):
    def __init__(self, albedo: Color):
        self.albedo = albedo

    def scatter(self, r_in: Ray, rec: HitRecord) -> tuple[bool, Color, Ray]:
        scatter_direction = rec.normal + random_unit_vector()

        # Catch degenerate scatter direction
        if scatter_direction.near_zero():
            scatter_direction = rec.normal

        scattered = Ray(rec.p, scatter_direction)
        return True, self.albedo, scattered


class Metal(Material):
    def __init__(self, albedo: Color, fuzz=0.0):
        self.albedo = albedo
        self.fuzz = min(fuzz, 1.0)

    def scatter(self, r_in: Ray, rec: HitRecord) -> tuple[bool, Color, Ray]:
        reflected = reflect(r_in.direction, rec.normal)
        scattered = Ray(rec.p, unit_vector(reflected) + random_unit_vector() * self.fuzz)
        scatter_happened = dot(scattered.direction, rec.normal) > 0
        return scatter_happened, self.albedo, scattered


class Dielectric(Material):
    def __init__(self, index_of_refraction: float):
        self.ir = index_of_refraction

    def reflectance(self, cosine: float, ref_idx: float) -> float:
        """Use Schlick's approximation for reflectance"""
        r0 = (1.0 - ref_idx) / (1.0 + ref_idx)
        r0 = r0 * r0
        return r0 + (1.0 - r0) * ((1.0 - cosine) ** 5)

    def scatter(self, r_in: Ray, rec: HitRecord) -> tuple[bool, Color, Ray]:
        attenuation = Color(1.0, 1.0, 1.0)
        refraction_ratio = 1.0 / self.ir if rec.front_face else self.ir

        unit_direction = unit_vector(r_in.direction)
        cos_theta = min(dot(-unit_direction, rec.normal), 1.0)
        sin_theta = math.sqrt(1.0 - cos_theta * cos_theta)

        cannot_refract = refraction_ratio * sin_theta > 1.0
        will_reflect = self.reflectance(cos_theta, refraction_ratio) > random.random()

        if cannot_refract or will_reflect:
            direction = reflect(unit_direction, rec.normal)
        else:
            direction = refract(unit_direction, rec.normal, refraction_ratio)

        scattered = Ray(rec.p, direction)
        return True, attenuation, scattered
